feat: add support for ResultSet#getObject(OffsetTime.class) and PreparedStatement#setObject(OffsetTime.class)

uschindler opened the pull request:
After #2464 I took care about the remaining conversions between Postgres wire format, legacy calendars and Timestamp/Date/Time and the modern java.time APIs.

This PR especially takes care of TIMETZ, which was converted from/to calendars and `java.sql.Time`. This implements a parser for the `TIMETZ` wire format and also allows them to get using `ResultSet#getObject(col, OffsetTime.class)`, which preserves the original server-stored timezone!

This also fixes `TimestampUtils#parseBackendTimestamp()` to use a `ZoneOffset` instead of allocating and modifying legacy calendar instances. This allows to easier read also the old legacy code but also the wire format parsers, because you don't need to calculate with raw offsets. This also removes state from `TimestampUtils`.

### All Submissions:

* [X] Have you followed the guidelines in our [Contributing](https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md) document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

1. [X] Does your submission pass tests?
2. [X] Does `./gradlew autostyleCheck checkstyleAll` pass ?
3. [X] Have you added your new test classes to an existing test suite in alphabetical order?

uschindler commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java:
```diff
@@ -671,6 +672,8 @@ public int getConcurrency() throws SQLException {
 
   }
 
+  private static final LocalDate LOCAL_DATE_EPOCH = LocalDate.of(1970, 1, 1);
```
There is a constant in later Java versions, but not Java 8

vlsi replied:
This should be better as a `// TODO` comment in code :)

uschindler replied:
+1

uschindler replied:
done.

uschindler commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java:
```diff
@@ -3740,13 +3750,20 @@ public void updateArray(String columnName, @Nullable Array x) throws SQLExceptio
       }
     } else if (type == OffsetDateTime.class) {
-      if (sqlType == Types.TIMESTAMP_WITH_TIMEZONE || sqlType == Types.TIMESTAMP) {
+      if (sqlType == Types.TIMESTAMP_WITH_TIMEZONE || sqlType == Types.TIMESTAMP || sqlType == Types.TIME) {
```
This was actually a bug (new test was added): https://github.com/pgjdbc/pgjdbc/pull/2467/files#diff-7e93771092eab9084402e3c7c81319a1f037febdc7614264329bd29f11d39ef2R690 says it must be supported

uschindler commented on pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java:
```diff
@@ -183,7 +153,8 @@ private Calendar getCalendar(int sign, int hr, int min, int sec) {
     int nanos = 0;
 
-    @Nullable Calendar tz = null;
+    boolean hasOffset = false;
```
this was added to get away with null/non-null checks. For legacy APIs where calendar is used, this is the signal to use default calendar

uschindler commented on pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java:
```diff
@@ -438,12 +408,64 @@ private ParsedTimestamp parseBackendTimestamp(String str) throws SQLException {
+    }
+
+    // question: use default timezone with offset of current day here?
+    ZoneOffset timeOffset = ZoneOffset.UTC;
```
The question here is: Should we use UTC or the local zone offset. The problem is without a date a timezone is useless, so I think UTC is correct here. Tis also matches OffsetDateTime behaviour

vlsi replied:
I see how "just use UTC" might sound good, however, I wonder if `time -> OffsetTime` conversion should be allowed at all.

What if we just throw an exception like "please use `getObject(.., LocalTime.class)` instead"?

uschindler replied:
This is exactly what I asked myself!
Currently it is possible also to get a `OffsetDateTime` also for `TIMESTAMP`, although this has the same (semantic) problem. Of course `TIMESTAMPTZ` in Postgres is always stored and returned in UTC (so actually both are the same on wire), so here it is fine to set always UTC as zone offset.
`TIMETZ` is special, because actually here Postgres saves the timezone. And also here we have the other assymmetry: You can do `getObject(LocalTime.class)` on `TIMETZ`. If we disallow getting `OffsetTime` from `TIME`, we must also disallow getting `LocalTime` from `TIMETZ`.

Actually at several places in code it mentions the JDBC 4.2 standard that seems to specify how the java.time classes are mapped to SQL types, so we should do what they say. I tried to download it, but it gave me just the API, but no spec about how types map. Do you have a copy of this PDF to check that is mentioned in several comments?

uschindler replied:
> You can do getObject(LocalTime.class) on TIMETZ. If we disallow getting OffsetTime from TIME, we must also disallow getting LocalTime from TIMETZ.

In current code this fails in binary representation, but works in text transmission. A test is clearly missing here!

So in my opinion, we should explicitely forbid both. What do you think?

uschindler commented on pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java:
```diff
@@ -1110,7 +1119,7 @@ public LocalTime toLocalTimeBin(byte[] bytes) throws PSQLException {
     }
 
-    return LocalTime.ofNanoOfDay(micros * 1000);
+    return LocalTime.ofNanoOfDay(Math.multiplyExact(micros, 1000L));
```
I adapted this like in my own code as it's better code style

uschindler replied:
Sorry it is not really "code style", it is more about safety. All java.time code uses multiplyExact to prevent overflow errors, so we should do this here, too. But if you disagree, I can undo this line of code.

davecramer replied:
I do not disagree with it, was just trying to understand the comment.

uschindler commented:
I just recognized, there's also no support for OffsetTime in PreparedStatement. I will add this in a moment. For completeness this should be added.

uschindler commented:
I had to revert to the shared calendar, although I moved to `ZoneOffset` in `TimestampUtils`. The reason for this is that the calendar cache is used if the time zone in parsed timestamp is *not* the local one.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc42/SetObject310Test.java:
```diff
@@ -160,7 +161,7 @@ private void insertWithoutType(Object data, String columnName) throws SQLExcepti
 
         assertTrue(rs.next());
-        return expectedType.cast(rs.getObject(1));
+        return type.cast(rs.getObject(1, type));
```
Why change this?
I think this was deliberately added to test `getObject(int)` behavior, and adding another argument changes the code path.

It is important that the driver does not accidentally change `getObject(int)` result even though there's no exact specification on what `getObject` should return.

uschindler replied:
This change actuall verifies that setObject/getObject is symmetric. I know that generic getObject() is not needed.

I will change this to actually do two tests to check both variants.
- The return as done here should be reverted, YES!
- Before retuning the generic Object value, the test should assert that "Object data" and the value returned by `rs.getObject(1, data.getClass()` are equal.

When doing this we check both (symmetry - which is way more important) and that the untyped `getObject()` returns the input unmodified.

uschindler replied:
I fixed this, see my last commit. The latest change brings both!

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc42/SetObject310Test.java:
```diff
@@ -185,7 +186,7 @@ private void insertWithoutType(Object data, String columnName) throws SQLExcepti
 
         assertTrue(rs.next());
-        return expectedType.cast(rs.getObject(1));
+        return type.cast(rs.getObject(1, type));
```
I suggest reverting the change a well.

uschindler replied:
will do this like mentioned before

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc42/TimestampUtilsTest.java:
```diff
@@ -62,6 +63,46 @@ public void testToLocalTime() throws SQLException {
+  public void testToStringOfOffsetTime() {
+    TimestampUtils timestampUtils = createTimestampUtils();
+
+    assertEquals("00:00:00+00", timestampUtils.toString(OffsetTime.parse("00:00:00+00:00")));
```
Would you please factor it into a `assertOffsetTimeToString(String offsetTime)` method?

Then the test code would be easier to read (less `timestampUtils.toString(OffsetTime.parse` repetitions), and the assertion messages would be easier to understand: the failure message could include the reason of the failure so CI failures are easier to understand.

For instance:


```java
void assertOffsetTimeToString(String inputTime, String expectedOutput) {
  // Note how failure message would include **input** value, so it would be easier to understand what went wrong.
  assertEquals(
    expectedOutput,
    timestampUtils.toString(OffsetTime.parse(inputTime)),
    () -> "timestampUtils.toString(OffsetTime.parse(" + inputTime + "))"
  );
}
```

An alternative option is to use `@ParameterizedTest`, however, most of our tests are still in JUnit4, so adding a method seems to be the way to go.

uschindler replied:
This is again copypaste unchanged from the corresponding LocalTime test. I did not want to change the pattern of other tests in that class.
Your ideas are good, but that should also be applied to the other tests in that class. But that's out of scope of this issue. Having all of them working the same way makes them easier to change later to use parameters.

uschindler replied:
Do we agree to not change this here and better make a cleanup of the whole `TimestampUtilsTest` class? I would like to resolve this here and won't fix.

uschindler replied:
I would not use parameters here, lets just add a method for the verification as suggested. Will do this also for LocalTime!

uschindler replied:
Hi, I rewrote the test with the pattern provided above (a separate assertion method). Because of this the comments about nanos are not needed because message says what is expected

vlsi replied:
The comments regarding "499 vs 500" would still be helpful for the edge cases.
It might make sense to have an overload like `assertToStringOfLocalTimeUnchanged(String expectedOutput) { assertToStringOfLocalTime(expectedOutput, expectedOutput);`

uschindler replied:
yes, i will readd the comments. I was thinking about adding them to the error message, but that seemed too much. I just have to revert.

Adding an overload is a good idea, not sure if this is really needed here. But it would bring the attention of reader directly to the "special" cases.

uschindler replied:
I won't add the overloads. Although you have duplicated strings the test looks more clear to me in its current state.

Comments were added back.

uschindler replied:
The other reason why I don't want to add this "unchanged": The formats by postgres and Java have a different definition of the formatting. Postgres uses a shorter zoneoffset notation. We actually want to test here how a LocalTime is converted to the postgres representation. That some of the results are identical to LocalTime.toString() is just happening but no requirement. For OffsetTime that's more obvious because of different timezone notation.

vlsi commented on pgjdbc/src/test/java/org/postgresql/test/jdbc42/TimestampUtilsTest.java:
```diff
@@ -62,6 +63,46 @@ public void testToLocalTime() throws SQLException {
+    assertEquals("23:59:59+01", timestampUtils.toString(OffsetTime.parse("23:59:59+01:00")));
+    assertEquals("23:59:59.999999+01", timestampUtils.toString(OffsetTime.parse("23:59:59.999999+01:00"))); // 0 NanoSeconds
+    assertEquals("23:59:59.999999+01", timestampUtils.toString(OffsetTime.parse("23:59:59.999999499+01:00"))); // 499 NanoSeconds
+    assertEquals("24:00:00+01", timestampUtils.toString(OffsetTime.parse("23:59:59.999999500+01:00")));// 500 NanoSeconds
```
`500 NanoSeconds` should be better as an extra message parameter to the assertion.
Code comment is visible only to those who read the code, and assertion message is visible when the test fails.

uschindler replied:
Same like before: i just duplicated already existing test and extended by timezones. The other test has same formatting.

uschindler replied:
I can move this to the extra message param. Will do the same for the other tests in the class or should I leave it as-is to not make this PR larger?

vlsi replied:
> should I leave it as-is to not make this PR larger?

That is up to you.
I would be ok if the old tests were left as is (e.g. to prove backward behaviour is intact)
I would be ok if the old tests were refactored to the newer approach.

I do not like adding new tests with "old" style because it makes maintenance harder. For instance, if a test fails, then it would be harder to understand why did it fail, and why a certain value was expected.

You add tests, and it is awesome. My primary ask here is to keep "newly added tests" easy to maintain (==ensure test failures look like a decent bug report).
So if you decide between "matching the old style" and "making the test comprehendable", I would almost always drop "old style", and write test in a way to make the assertion failure easy to reason.

uschindler replied:
I fixed this test to have the rounding conditions in the message. I did this also for LocalTime.

vlsi commented on pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java:
```diff
@@ -110,7 +110,7 @@
         tzField.setAccessible(true);
         TimeZone defaultTz = TimeZone.getDefault();
-        @SuppressWarnings("nulllability")
+        @SuppressWarnings("nullability")
```
It is not needed

uschindler commented:
I changed the `GetObject310Test` after @vlsi's comment to check both:
- `getObject()` without type returns the native type and checks if it is the expected one. For the new test of OffsetTime the return value is not explicitely tested. This restores the old behaviour before the PR.
- it also tests roundtrip of `setObject/getObject` with type. This is a really good test and checks the symmetry. Nice side-effect: It is applied to all existing tests in the class!

vlsi commented:
@sehrope , @bokken , @marschall , @davecramer , do you have opinions/ideas if the driver should throw errors if user asks `getObject(int, OffsetTime.class)` for `time withOUT time zone` and `getObject(int, OffsetDateTime.class)` for `timestamp withOUT time zone`?

I am inclined those API calls should be forbidden. At the end of the day, the developer should use proper data types. If they know they use `timestamp` (without timezone)  at the database, then they should just use `LocalDateTime.class`  rather than `OffsetDateTime.class`.

As a middle-ground, there might be a connection property like `what_to_do_when_reading_local_datetime_as_timezoned=throw_error (default) | use_utc | use_default_timezone`.

WDYT?

sehrope commented:
@vlsi  This is brand new functionality right? i.e. previously asking `getObject(...)` for an OffsetDateTime wouldn't have worked at all so we have the luxury of doing things correctly. If so, then yes it should throw an error if used with `timestamp withOUT time zone`.

-1 to adding a flag to allow the behavior. We have too many flags that change handling already. And anybody that can configure a flag for a new application using this data type can change their code to use the right type.

vlsi commented:
Unfortunately, the current code does have silent UTC injection for TIMESTAMP -> OffsetDateTime conversion: https://github.com/pgjdbc/pgjdbc/blob/c02aa7725a3c5c71314e0e8c441a72fc97ff83a0/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L684-L685 and https://github.com/pgjdbc/pgjdbc/blob/c02aa7725a3c5c71314e0e8c441a72fc97ff83a0/pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java#L553-L556.

uschindler commented:
Hi,
@sehrope this is partly new functionality, but the already existing code is partly not working at all or silently introducing UTC timezone (e.g when you request a `OffsetDateTime` from a "timestamp without tz" column).

See the full discussion here: https://github.com/pgjdbc/pgjdbc/pull/2467#discussion_r821831584

This PR mostly adds support for reading the `OffsetTime` wire format natively, because as mentioned in the previous PR the current code converts between legacy `java.sql.Time` and new `java.time` APIs in non-predictable ways (depending on default timezone).

While fixing this single case (TIMETZ -> OffsetDateTime), I had to add the TIMETZ wire format parser, which is a very good thing, because it allows you to get the full TIMETZ including all information and you don't break. This is the only addition by the PR, but the problem existed before.

While implementing this and adding tests I also noticed more bugs in current code (e.g., reading LocalTime from a TIMETZ column works with string wire format, but fails with binary wire format). This is still not fixed in this PR, but I'd like to do this, too (it's all messy, I hate this long chains of instanceof and if-checks in the call hierarchy - there's no common place where the input/output formats are validated). At least we should be consistent between string and binary wire format. Because of this there should be a decission (optimally based on the JDBC 4.2 standard that I don't have) how to handle conversions between non-TZ and TZ fomats with `java.time` APIs.

uschindler commented:
> `getObject(...)` for an `OffsetDateTime`

this worked before. New is `OffsetTime`, but both have same problem!

uschindler commented:
> So we have the luxury to make `getObject(..., OffsetTime.class)` to throw an error when column type is "local time". However, I would like to move `OffsetDateTime` in the direction of `throw when column type is known to be local timestamp`.
> 
> Of course, `OffsetDateTime` is not relevant for this PR, and we probably need to focus on `time` -> `OffsetTime` conversion. It is new. Should it just throw?
> 
> The next fun idea might be gradual movement for `getObject(int)` to `java.time.` classes instead of ambiguous `java.sql.Timestamp`

I think this is not allowed according to the spec. I would like to do this, too, but this would heavily break user code.

What we should really fix after we have Java 8 as minimum is also maintain the mapping of the new Constants in `java.sql.Types`. Currently the ResultSetMetadata returns `TIME` or `TIMESTAMP` always.

uschindler commented:
> > The next fun idea might be gradual movement for `getObject(int)` to `java.time.` classes instead of ambiguous `java.sql.Timestamp`
> 
> I think this is not allowed according to the spec. I would like to do this, too, but this would heavily break user code.
> 
> What we should really fix because we have Java 8 as minimum is also maintain the mapping of the new Constants in `java.sql.Types`. Currently the ResultSetMetadata returns `TIME` or `TIMESTAMP` always.

See, TABLE B.3 in first PDF here: https://download.oracle.com/otndocs/jcp/jdbc-4_2-mrel2-spec/index.html

This also contains the information I asked for above: JDBC 4.2 requires to support LocalDate, LocalTime, LocalDateTime, OffsetTime, OffsetDateTime for getObject(see table B.4/B.5, unfortunately it does not describe how conversions should be done).

uschindler commented:
...but if you read page 197 for the setObject it lists all datatypes that you can convert on setObject(). Here it lists the following conversions from java.time types to SQL types on setObject:

- java.time.LocalDate: CHAR, VARCHAR, LONGVARCHAR, DATE
- java.time.LocalTime: CHAR, VARCHAR, LONGVARCHAR, TIME
- java.time.LocalDateTime: CHAR, VARCHAR, LONGVARCHAR, DATE, TIME, TIMESTAMP
- java.time.OffsetTime: CHAR, VARCHAR, LONGVARCHAR, TIME_WITH_TIMEZONE
- java.time.OffsetDatetime: CHAR, VARCHAR, LONGVARCHAR, TIME_WITH_TIMEZONE, TIMESTAMP_WITH_TIMEZONE

For the inverse transformation it has no information, but the above list clearly specifies that on setObject you should not be able to set a OffsetTime on TIME and LocalTime on TIMETZ!

uschindler commented:
Another option would be to log a warning to SQLLogger?

vlsi commented:
@uschindler , thank you for digging into this. I like how the spec allows throwing errors, so users could see they use wrong APIs.

>Another option would be to log a warning to (SQL/Connection)Logger?

I would like to try gradual deprecation here.

add a property like `what_to_do_when_reading_local_datetime_as_timezoned=print_warning (default) | throw_error | use_utc | use_default_timezone`

Step 1: log warning, allow users to try new behavior
Step 2: flip default and still allow to configure old behavior if it is still important
Step 3 (never?): remove old behavior from code.

In this case, the added complexity sounds minimal (we either throw an error or we add UTC timezone).
Testing both values would be harder than testing just one of them, however, I see no other way to flip the default.

uschindler commented:
I don't really like the complexity and adding an extra parameter, but I can see what I can do. I may need help with implementing this as I don't know how the connection parameters are available from TimestampUtils / PGResultSet /PGPreparedStatemen.

Maybe proceed as follows:
- for OffsetTime don't allow mixing timezone and no timezone. The code in TimestampUtils to read a binary timestamp no longer accespt `bytes.length!=12`
- for OffsetDateTime I will for now keep the code but also make it consistent between text/binary transfer and just log a warning if you try to get a OffsetDateTime from a plain timestamp. I will keep setting it to UTC.

The propsed changes here could maybe done by somebody having better knowledge.

uschindler commented:
I pushed code to implement this:
- Disallow OffsetTime on non-TZ times.
- Also check the OID on string types, previously in string mode you were able to read a TIME from a TIMESTAMPTZ, in binary mode this was not possible.

uschindler commented:
Personally I like the new code for the java.time getters. The code picks the cases it allows (all OID types) and returns the converted value. At end of method I added the type conversion error, so all code path at end throw conversion exception.  In Apache Lucene we use that pattern quite often, maybe its useful here, too.

uschindler commented:
I try to cleanup the type checks splattered over the whole PGResultSet. 

The checks here are obsolete, because the methods "get(Local|Offset)(Date|Time|DateTime)" are checking types already: https://github.com/pgjdbc/pgjdbc/blob/36c6a001e6346dc2090b931e9244ea91cd9340d3/pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java#L3743-L3779

Any opinions on removing them?

vlsi commented:
>Any opinions on removing them?

+1 for removing checks in `getObject`.

uschindler commented:
Removed the type checks.

uschindler commented:
To me this PR looks like covering all problems. About the warnings whenconverting timestamps without TZ to OffsetDateTime we should open a separate issue.

The only thing I will add after lunch is a few tests on `GetObject310Test` to check that wrong type casts are prevented in all modes (binary, string). This is easy by adding just a "false get" with assertThrows.

uschindler commented:
My final word about the OffsetDateTime: We should maybe really only log a warning, because for timestamp vs timestamptz theres not really an issue behind the scenes: timestamp and timestamptz are in Postgres the exact same on the wire format, because the timestamps are always converted to UTC before saving them in database. When you retrieve the values as string you will get back a timezone relative to server's default, which OffsetDateTime read code converts to UTC (to make it consistent with the binary transfer, otherwise it could happen that one suddenly gets UTC when the server switches to binary transfer after the compile threshhold).

In short: Nothing really fails for somebody when an OffsetDateTime is retrieved from a timestamp or timestamptz, it is always UTC.

davecramer commented:
> @sehrope , @bokken , @marschall , @davecramer , do you have opinions/ideas if the driver should throw errors if user asks `getObject(int, OffsetTime.class)` for `time withOUT time zone` and `getObject(int, OffsetDateTime.class)` for `timestamp withOUT time zone`?
> 
> I am inclined those API calls should be forbidden. At the end of the day, the developer should use proper data types. If they know they use `timestamp` (without timezone) at the database, then they should just use `LocalDateTime.class` rather than `OffsetDateTime.class`.
> 
> As a middle-ground, there might be a connection property like `what_to_do_when_reading_local_datetime_as_timezoned=throw_error (default) | use_utc | use_default_timezone`.
> 
> WDYT?

Sorry for the late reply

I think we should add the connection property and if throwing an error is new behaviour then we should use the default that enables the old behaviour. If people want strict compliance they can request it.

Dave

uschindler commented on pgjdbc/src/test/java/org/postgresql/test/jdbc42/GetObject310Test.java:
```diff
@@ -287,12 +314,12 @@ public void localTimestamps(ZoneId zoneId, String timestamp) throws SQLException
+
+        assertDataTypeMismatch(rs, "timestamp_without_time_zone_column", OffsetTime.class);
+        // TODO: this should also not work, but that's an open discussion (see https://github.com/pgjdbc/pgjdbc/pull/2467):
+        // assertDataTypeMismatch(rs, "timestamp_without_time_zone_column", OffsetDateTime.class);
```
This is the only conversion which was allowed by previous code. I am not sure if it really makes sense to add a sepcial case with a connectuion parameter. I would either:
- change this and disallow this type conversion (I don't expect anybody using this, if yes: they could quickly change)
- log a warning. If somebody is annoyed by the warning he/se can also change code

uschindler replied:
@davecramer this is the test which I added in this PR mentioned in my reply: In an ideal world it should be enabled. But this would require a change in behaviour.

davecramer replied:
Ah, ok, this is what I didn't understand. Thanks for the clarification

uschindler replied:
The new code with OffsetTime does not accept wrong conversions (as it is completely new), the old code with OffsetDateTime still does work with timestamp (without TZ).

Thanks for review!

uschindler commented:
Hi @davecramer, hi @vlsi, 
thanks for comments. I added tests for all illegal type conversions (see latest commit). There is only one conversion which was allowed by the old code - marked by a TODO. As said before, I don't think it is worth to add a sepacial case with a connection parameter! I would either change this to behave correct or log a warning. Anybody annoyed by the warning can change the code and think about it.

Sorry for the delay, I forgot to add the tests. I also cleaned up the whole GetObject310 test and added try-with-resources everywhere. I also replaced the functional interfaces by lambdas.

vlsi commented:
>I don't think it is worth to add a sepacial case with a connection parameter!

Our experience with pgjdbc is that many people use it, and some of them use the driver in extremely unexpected ways, so changing behaviours (including fixing bugs) can easily become painful for the consumers since "you never know if the upgrade is safe".

Then, date-time-timestamp are hard to get right, so I can easily imagine someone might already implement their workarounds on top of ill conversions within the driver.
That is why I think we can't just remove ill conversion, and the only way we can remove it is to add a property, and then flip the property.

In general, I do not like adding more and more properties since it makes testing harder, however, it looks like the property is the only viable way to deprecate and de-support illegal conversion which was allowed by the old code.

----

All in all, adding a property is a different story, and it can be a part of a different PR.
I just explain the reason I don't like to drop old illegal conversion right now.

uschindler commented:
> All in all, adding a property is a different story, and it can be a part of a different PR.
> I just explain the reason I don't like to drop old illegal conversion right now.

Thant's OK. The "illegal conversion" was just uncovered by my code review and testing, but is unrelated to this issue, which adds OffsetTime support and removes the Database -> legacy sql.Time -> OffsetTime roundtrip for one of the conversions.

Another PR is fine. My plan was only to add a warning so the wrong usage gets verbose. This wouldn't have breaked any code out there just made it noisy.

Anyways, I am one with this one, so it is free for review.

davecramer commented:
@uschindler thanks for all this work I will try to review it this weekend. I'd like to release soon, so I'd like to get this in.

@sehrope ? @bokken ? @jorsol ?

uschindler commented:
> So I'm OK with this PR with one provision. If we do not add a connection parameter then this will have to go into 42.4.x if we add a connection parameter it can go into 42.3.x My preference is a connection parameter and put it into 42.3.x

This PR does NOT yet change anything in behaviour regarding the TZ vs non-TZ types, see the TODO: https://github.com/pgjdbc/pgjdbc/pull/2467/files#diff-7e93771092eab9084402e3c7c81319a1f037febdc7614264329bd29f11d39ef2R664

Maybe you misinterpreted the discussion above.

davecramer commented:
> > So I'm OK with this PR with one provision. If we do not add a connection parameter then this will have to go into 42.4.x if we add a connection parameter it can go into 42.3.x My preference is a connection parameter and put it into 42.3.x
> 
> This PR does NOT yet change anything in behaviour regarding the TZ vs non-TZ types, see the TODO: https://github.com/pgjdbc/pgjdbc/pull/2467/files#diff-7e93771092eab9084402e3c7c81319a1f037febdc7614264329bd29f11d39ef2R664
> 
> If we disallow the conversion as noted in above TODO (and enable the test, I commented out), then it changes behaviour!
> 
> Maybe you misinterpreted the discussion above.

Apparently so.

Thanks for the clarification

davecramer commented:
I'm going to merge this unless there are objections ?

vlsi commented on pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java:
```diff
@@ -719,18 +713,19 @@ public int getConcurrency() throws SQLException {
     }
 
-    String string = castNonNull(getString(i));
-    return getTimestampUtils().toLocalDateTime(string);
+    throw new PSQLException(
+        GT.tr("Cannot convert the column of type {0} to requested type {1}.",
+            Oid.toString(oid), "java.time.LocalDateTime"),
+        PSQLState.DATA_TYPE_MISMATCH);
```
I would slightly disagree with moving throw at the end of the method.
If we require `Oid.TIMESTAMP`, then we just throw at the very beginning, so it is clear what are the valid input values, and we reduce the nesting of the subsequent code.

uschindler replied:
I can check this again. I made the code here consistent with the other similar methods where the logic to check is more complicated.

For this part I can reverse it, but for the other methods where more than one type is expected, this would require to duplicate checks.

I think this is a matter of code style, I prefer it like that because its easier readable.

vlsi commented on pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java:
```diff
@@ -143,30 +143,13 @@ public TimestampUtils(boolean usesDouble, Provider<TimeZone> timeZoneProvider) {
-    if (sec < 10) {
-      zoneID.append('0');
-    }
-    zoneID.append(sec);
-
-    TimeZone syntheticTZ = new SimpleTimeZone(rawOffset, zoneID.toString());
-    calCache = new GregorianCalendar(syntheticTZ);
-    calCacheZone = rawOffset;
+    calCache = new GregorianCalendar(TimeZone.getTimeZone(offset));
+    calCacheZone = offset;
```
I wonder if this change has performance implications.
`ZoneOffset` should be fine as it stores the offset via `int totalSeconds`.

@uschindler , have you checked `TimeZone.getTimeZone(offset)` performance?

uschindler replied:
I did not do a performance check, but `TimeZone.getTimeZone(offset)` is not globally synchronized like the one taking a string parameter.

There was a previous commit where this was changed at another place: https://github.com/pgjdbc/pgjdbc/commit/beeabe2ead05e0cc6303acd6e3ff65a26623d2fd

I will come back to this a bit later by checking what going on behind the scenes.

uschindler replied:
I checked the code in `TimeZone.getTimeZone(ZoneOffset)` (Java 8): Basically it is doing more or less the same. It 

If you like, you can replace the code and also use `SimpleTimeZone` with `"GMT" + offset.getId()` and call it like before, but `TimeZone.getTimeZone(ZoneOffset)` has caching for the zone ids, so it looks better to use this code.

At least there's no synchronization involved.

I am not sure whet the best way to check performance is, you are the better specialist!

uschindler replied:
In addition, there's also the cache, so the performance of this method is not significant (unless the strings to be parsed all the time have different time zones). This is unlikely because the backend (Postgres) returns the same zone offsets (server's local one, but only in text-representations). After some time the JDBC driver switches to binary transmission and then this method is no longer called.

So I don't think this is an issue. The cache is helpful to work around the raw-offset to string conversions.

uschindler replied:
If you still feel better, I can change the method to:

```java
  private Calendar getCalendar(ZoneOffset offset) {
    if (calCache != null && Objects.equals(offset, calCacheZone)) {
      return calCache;
    }

    final String tzid = (offset.getTotalSeconds() == 0) ? "UTC" : "GMT".concat(offset.getId());
    final TimeZone syntheticTZ = new SimpleTimeZone(offset.getTotalSeconds() * 1000, tzid);
    calCache = new GregorianCalendar(syntheticTZ);
    calCacheZone = offset;
    return calCache;
  }
```

The `tzid` string operations are needed because the ID of raw offsets are not using "GMT" prefix and in case of UTC, it returns "Z". But actually that doe snot matter, we could also add a constant fake ID, because the ID is never used by Java or Calendar.

uschindler replied:
The conversions of the name were adapted from the conversions original `TimeZone.getTimeZone(ZoneId)` source code.

uschindler replied:
Hi, we figured out that there is some strange difference between `SimpleTimeZone(offset.getTotalSeconds() * 1000)` and `TimeZone.getTimeZone(offset)`.

So we should fix this in #2488 by applying above suggestion. This is still a server-side problem in combination with problematic Java time zone handling for offsets going down to minutes/seconds.

uschindler replied:
Here is the JDK bug report: https://bugs.openjdk.java.net/browse/JDK-8285844

davecramer commented:
I'd really like to release this week so I will be merging this tomorrow unless someone objects
@vlsi we can tweak the code a bit after it is merged

uschindler commented:
Thanks for merging!

uschindler commented on pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java:
```diff
@@ -403,7 +386,7 @@ private ParsedTimestamp parseBackendTimestamp(String str) throws SQLException {
 
     ParsedTimestamp ts = parseBackendTimestamp(s);
-    Calendar useCal = ts.tz != null ? ts.tz : setupCalendar(cal);
+    Calendar useCal = ts.hasOffset ? getCalendar(ts.offset) : setupCalendar(cal);
```
Hi Dave,
what regression do you mean, because this change (at öeats this line is not anything that could break the parsing). The code change here is required to get rid of java.util.Calendar when just parsing the datestamp to ParsedTimestamp. In the previous code `ParsedTimestamp#tz` was a `Calendar` instance and not a plain timezone offset. For parsing a String date/time to an `OffsetTime` the calendar is not helpful. 

So the refactoring was needed to simple save the raw time offset in ParsedTimestamp and mve the code using the calendar to somewhere else. If there is a bug in it it would be good to see an example of the regression you are seing.

uschindler replied:
Are you referring to this comment? https://github.com/pgjdbc/pgjdbc/pull/2476#discussion_r848790128

uschindler replied:
From what I understand the issue in #2476 is a different one. It is about LocalDate and BCE. @vlsi mentioned that we want to possibly fix it. I don't think this has anything to do with this PR, it may be caused by the previous one (#2464)?

uschindler replied:
I think we should add a test that fails and open a separate PR to fix a possible year shift for BC LocalDate (caused by a different defintiion of Year 0 in Postgresql and Java).

uschindler replied:
When you give me an hint about the exact problem, I can figure out if we maybe should fix this like proposed here: https://github.com/pgjdbc/pgjdbc/pull/2467/files/918926e6e92789381058dd8b1ea2946f2329dc70#r833415824
